### PR TITLE
Update with pub sub replication and Redis driver

### PR DESCRIPTION
**Pull request by snellingio**
Still needing tests, and handling edge cases such as statistics, etc.

### CODE REVIEW

**Comment by snellingio** [comment]
This references #6

**Comment by snellingio** [comment]
Oh, and since this is my PR, my biggest feature request for this is figuring out how to use the cli logger in the class. I'd like to have a lot more logging in the class. We can easily do this with the `->then()` function within the redis class. But adding the logging in a clean way was where I spent a little too much time over the weekend, and then ripped it out.

File: PubSub/Redis/RedisClient.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        $payload['appId']    = $appId;
+        $payload['serverId'] = $this->serverId;
+        $this->publishClient->publish(self::REDIS_KEY, json_encode($payload));
+        return true;
+    }
+
+    public function subscribe(LoopInterface $loop): PubSubInterface
+    {
+        $this->loop = $loop;
+        [$this->publishClient, $this->subscribeClient] = Block\awaitAll([$this->publishConnection(), $this->subscribeConnection()], $this->loop);
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply PubSub/Redis/RedisClient.php]: To be honest I'm more wondering why you are blocking here. Is this bit ran inside the websocket server, or somewhere outside it in blocking PHP land?

[the remark this answers is not in the document]
Comment (snellingio) [reply PubSub/Redis/RedisClient.php]: If you don't await, you will end up returning a promise. Which isn't okay if you want the publish client to end up in the container. So what we do here is block until the promises resolve on the loop, and then push the actual instances on the variables instead of promises. Hopefully that makes sense.

[the remark this answers is not in the document]
Comment (snellingio) [reply PubSub/Redis/RedisClient.php]: So this is going to be slightly confusing, because I don't fully understand Ratchet. But my understanding is that there is an event loop that is always running that you can push processes onto. The loop is started from the console command here: https://github.com/beyondcode/laravel-websockets/blob/c1f6ffa51b60a896d44fff490f1f6631b1758c53/src/Console/StartWebSocketServer.php#L36

You don't need to restart the loop. Just block for a split second to get the redis connection promise to resolve, and then the loop continues.

Since this subscribe command only runs once, it will just connect when you boot up the websocket server, and then continue along in the background.

[the remark this answers is not in the document]
Comment (snellingio) [reply PubSub/Redis/RedisClient.php]: That DI waiting client is a very interesting way to do it, and I'm totally open to it.

[the remark this answers is not in the document]
Comment (snellingio) [reply PubSub/Redis/RedisClient.php]: Sorry for the triple comment, but one other thing to consider will be implementing the interface in another client such as zeromq or some other pub/sub framework.

[the remark this answers is not in the document]
Comment (WyriHaximus) [reply PubSub/Redis/RedisClient.php]: No worries I'm on the ReactPHP core team. Anyway in short: the event loop is the core of any ReactPHP application, on which Ratchet, on which this package is build. And the loop is essentially a: 

```php
while (true) {
    if ($stopped === true) {
        break;
    }
    // Do stuff
}
```

Once started it will started it will continue running until there is nothing left to do. You start it with `$loop->run()` and stop it with `$loop->stop()`. Once stopped it won't start on it's own again unless you call `$loop->run()` again later in the code. (And if you haven't reset it everything is `pauzed`. (Not advisable because connections are waiting and will time out at some point.)) Now the `await*` functions will call `$loop->stop()` once the promises resolve/reject. Hence my questions and my guess your redis connections are intiated before the server starts. But assuming (and you should) the one loop (to rule them all) is shared all across the application so everything hooks into the same loop. (Using more then one loop leads to unforeseen complications and problems.)

> That DI waiting client is a very interesting way to do it, and I'm totally open to it.

Cool I'll put it in a separate package tomorrow.

> Sorry for the triple comment, but one other thing to consider will be implementing the interface in another client such as zeromq or some other pub/sub framework.

Cool, but you probably will run into the same return type issue with promises again. Maybe this could/should be solved at the interface level and it's users.

[the remark this answers is not in the document]
Comment (snellingio) [reply PubSub/Redis/RedisClient.php]: Thanks for the clarification on the loop. Makes a lot more sense.

>Cool, but you probably will run into the same return type issue with promises again. 

Yeah I guess this is my only question. Using the Waiting Client seems like the perfect way to hook onto it as a subscriber, but how should the publisher be passed back?

[the remark this answers is not in the document]
Comment (WyriHaximus) [reply PubSub/Redis/RedisClient.php]: Waiting client stand alone package is up at https://packagist.org/packages/wyrihaximus/react-redis-waiting-client I'll be adding integration tests for subscribing to that over the next few days

**Comment by snellingio** [comment]
I will not be able to pick this back up until after the new year. While I'll be around to communicate, I just won't have the time to advance this pull request during the holidays. 

Hope everyone has a great holiday.

**Comment by francislavoie** [comment]
@snellingio any chance you'll be picking this back up soon? I can probably give you a hand in finishing this PR soon, I'm working on a new project which will be using this lib.

File: PubSub/Redis/RedisClient.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (francislavoie) [reply PubSub/Redis/RedisClient.php]: Just to clarify, because it's clear the blocking approach was taken from my comment here: https://github.com/simonhamp/laravel-echo-ratchet-server/issues/2#issue-312013545, the reason I did it that way initially I think is just to lower the startup time by doing both connections at once. It's not strictly necessary. And like @snellingio said, it only ever happens right before the server is started, so there's no concern of the loop stopping.

File: PubSub/Redis/RedisClient.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+        /* @var $channelManager ChannelManager */
+        $channelManager = app(ChannelManager::class);
+        $channelSearch        = $channelManager->find($payload->appId, $payload->channel);
+
+        if ($channelSearch === null) {
+            return false;
+        }
+
+        $channel->broadcast($payload);
```
[the remark this answers is not in the document]
Comment (francislavoie) [reply PubSub/Redis/RedisClient.php]: It's to support this feature: https://laravel.com/docs/5.7/broadcasting#broadcasting-events

If a socketId is provided in the payload, then `broadcastToEveryoneExcept` should be called. Essentially the point of this is for example in a chat client, a user sent out a message, there's no reason to send an event about that message back to the same user who just sent it, since the frontend already has that message in its state. It's just a minor optimization that makes it send N-1 messages instead of N (where N is number of connections in the channel)

**Comment by snellingio** [comment]
@francislavoie I'd love some help finishing the PR. Especially around tests for this change.

File: PubSub/Redis/RedisClient.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply PubSub/Redis/RedisClient.php]: > Waiting client stand alone package is up at [packagist.org/packages/wyrihaximus/react-redis-waiting-client](https://packagist.org/packages/wyrihaximus/react-redis-waiting-client) I'll be adding integration tests for subscribing to that over the next few days

A better alternative for that package is up: https://github.com/clue/reactphp-redis/pull/82

[the remark this answers is not in the document]
Comment (WyriHaximus) [reply PubSub/Redis/RedisClient.php]: The lazy client release is out. That should make some thing simpler: https://github.com/clue/reactphp-redis/releases/tag/v2.3.0

[the remark this answers is not in the document]
Comment (francislavoie) [reply PubSub/Redis/RedisClient.php]: Thanks @WyriHaximus 

I'll try to pick up this PR in a bit, might make a new PR based on this one to finish it up.

**Comment by snellingio** [comment]
Closing in favor of #140 by @francislavoie
